cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,282 @@
+  }
+
+  using MemoTable = typename HashTraits<Type>::MemoTableType;
+  std::unique_ptr<MemoTable> memo_table_;
+};
+
+// ----------------------------------------------------------------------
+// Check if value in both arrays or not and returns boolean values
+
+// \brief Iterate over the left array using another visitor.
+// In VisitValue, use the memo_table_ (for right array) and check if value
+// in left array is in the memo_table_. Append the true to the
+// BooleanBuilder if condition satisfied, else false.
+
+// Additional member "right_null_count" is used to check if
+// null count in right is not 0, then append true to the BooleanBuilder
+// when left array has a null, else append false.
+
+template <typename T, typename Scalar>
+class IsInKernel : public HashBinaryKernelImpl {
+ public:
+  IsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool) {}
+
+  Status VisitNull() {
+    if (right_null_count != 0) {
+      bool_builder_.UnsafeAppend(true);
+    } else {
+      bool_builder_.UnsafeAppend(false);
+    }
+    return Status::OK();
+  }
+
+  Status VisitValue(const Scalar& value) {
+    if (memo_table_->Get(value) != -1) {
+      bool_builder_.UnsafeAppend(true);
+    } else {
+      bool_builder_.UnsafeAppend(false);
+    }
+    return Status::OK();
+  }
+
+  Status Compute(FunctionContext* ctx, const ArrayData& left, const ArrayData& right,
+                 std::shared_ptr<ArrayData>& out) override {
+    bool_builder_.Reset();
+    bool_builder_.Reserve(left.length);
+    memo_table_.reset(new MemoTable(0));
+
+    MemoTableRight<T, Scalar> func;
+    func.Append(right);
+    memo_table_ = std::move(func.memo_table_);
+
+    right_null_count = right.null_count;
+    ArrayDataVisitor<T>::Visit(left, this);
+
+    RETURN_NOT_OK(bool_builder_.FinishInternal(&out));
+    return Status::OK();
+  }
+
+ protected:
+  using MemoTable = typename HashTraits<T>::MemoTableType;
+  std::unique_ptr<MemoTable> memo_table_;
+  int64_t right_null_count;
+
+ private:
+  BooleanBuilder bool_builder_;
+};
+
+// ----------------------------------------------------------------------
+// (NullType has a separate implementation)
+
+// \brief When array is NullType, based on the null_count for the arrays,
+// append true/false tp the BooleanBuilder
+
+class NullIsInKernel : public HashBinaryKernelImpl {
+ public:
+  NullIsInKernel(const std::shared_ptr<DataType>& type, MemoryPool* pool) {}
+
+  Status Compute(FunctionContext* ctx, const ArrayData& left, const ArrayData& right,
+                 std::shared_ptr<ArrayData>& out) override {
+    bool_builder_.Reset();
+    bool_builder_.Reserve(left.length);
+
+    if (left.null_count != 0 && right.null_count != 0) {
+      for (int64_t i = 0; i < left.length; ++i) {
+        bool_builder_.UnsafeAppend(true);
+      }
+    } else if (left.null_count != 0 && right.null_count == 0) {
+      for (int64_t i = 0; i < left.length; ++i) {
+        bool_builder_.UnsafeAppend(false);
+      }
+    }
```

Comment:
What about the other cases, left.null_count == 0? See `SetAllNulls` in util-internal.h

Reply:
For 0 nulls in left array, result will be empty array because there will be no match.
Example:
left : 0 nulls
right: 3 nulls
expected output: []
kernel output: []